# Data Wrangling Project - JSON
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

****
## 1. Find the 10 countries with most projects


In [15]:
import numpy as np
import pandas as pd
import json
from pandas.io.json import json_normalize

In [9]:
#
# Create the world bank dataframe
#
df = pd.read_json('data/world_bank_projects.json')

In [111]:
#
# Print the head.
#
df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   
2    {'Name': 'Regional integration', 'Percent': 46}   
3  {'Name': 'Participation and civic engagement',...   
4  {'Name': 'Export development and competitivene...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [28]:
#
# Obtain a list of the 10 countries with the most projects.
#
df.groupby(['countryname']).size().sort_values(ascending=False).head(10)

countryname
People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
dtype: int64

## 2. Find the top 10 major project themes (using column 'mjtheme_namecode')

In [105]:
#
# Load the JSON.
#
json_str = json.load((open('data/world_bank_projects.json')))

#
# Flatten the string into a dataframe on the 'mjtheme_namecode' field.
#
theme_df = json_normalize(json_str, 'mjtheme_namecode', ['id'])

#
# Print the head.
#
theme_df.head()

code                                   name       id
0    8                      Human development  P129828
1   11                                         P129828
2    1                    Economic management  P144674
3    6  Social protection and risk management  P144674
4    5                  Trade and integration  P145310

In [106]:
#
# Obtain the 10 countries with the most projects.
#
theme_df.name.value_counts().sort_values(ascending=False).head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

** As you can see above, 122 projects do not have a name string associated with it.**

## 3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [100]:
#
# Print the head.
#
theme_df.head()

code                                   name       id
0    8                      Human development  P129828
1   11                                         P129828
2    1                    Economic management  P144674
3    6  Social protection and risk management  P144674
4    5                  Trade and integration  P145310

In [107]:
#
# Drop the ID column.
#
theme_df = theme_df.drop('id', axis=1)

#
# Print the head.
#
theme_df.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

In [109]:
#
# Get unique combination of code and name from dataset without empty strings.
#
unique_df = theme_df[theme_df.name != ''].drop_duplicates()

#
# Take the codes from the original df and merge with the unique df on the 'code' column.
#
merged_df = pd.DataFrame(theme_df.code).merge(unique_df, on='code', how='left')

#
# Perform a recount of the 10 countries with the most projects.
#
merged_df.name.value_counts().sort_values(ascending=False).head(10)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: name, dtype: int64

** As you can see, the empty names have been replaced and the counts have been updated to reflect the new totals. **